synapse/lib/lmdbslab.py

```diff
@@ -166,6 +166,229 @@ def inc(self, name, valu=1):
         self.set(name, curv)
         return curv
 
+class SafeKeyVal:
+    '''
+    Key/value storage that does not store items in memory and safely handles large keys.
+    '''
+    def __init__(self, slab, name, prefix=None):
+
+        self.name = name
+        self.slab = slab
+        self.valudb = slab.initdb(name)
+        self.abrvdb = slab.initdb(f'{name}:abrv2key')
+
+        self.prefix = prefix
+        if prefix is not None:
+            self.preflen = len(prefix)
+
+        self.offs = 0
+
+        item = self.slab.last(db=self.abrvdb)
+        if item is not None:
+            self.offs = s_common.int64un(item[0]) + 1
+
+    def getSubKeyVal(self, prefix):
+
+        if self.prefix is not None:
+            prefix = self.prefix + prefix
+
+        return SafeKeyVal(self.slab, self.name, prefix=prefix)
+
+    def get(self, name, defv=None):
+        '''
+        Get the value for a key.
+
+        Note:
+            This may only be used for keys < 503 characters in length. Use aget()
+            for keys above this limit or of unknown length.
+        '''
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        lkey = name.encode('utf8')
+
+        if len(lkey) > 502:
+            raise s_exc.BadArg('SafeKeyVal.get() may only be used with keys < 503 characters in length.')
+
+        if (byts := self.slab.get(lkey, db=self.valudb)) is None:
+            return defv
+        return s_msgpack.un(byts)
+
+    async def aget(self, name, defv=None):
+        '''
+        Get the value for a key.
+
+        Note:
+            This may be used for keys of any length. The sync get() method may be
+            used for keys which have a known length < 503 characters.
+        '''
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        lkey = name.encode('utf8')
+
+        if len(lkey) <= 502:
+            if (byts := self.slab.get(lkey, db=self.valudb)) is None:
+                return defv
+            return s_msgpack.un(byts)
+
+        for (abrv, realkey) in self.slab.scanByPref(lkey[:502], db=self.abrvdb):
+            if lkey == realkey:
+                if (byts := self.slab.get(lkey, db=self.valudb)) is None:
+                    return defv
+                return s_msgpack.un(byts)
+
+            await asyncio.sleep(0)
```

Comment:
Is this method async for this loop ( and subsequent sleep) just to handle 502 byte string prefix collisions?

Reply:
Yeah there is a sync `get` method for keys known to be <= 502 bytes and this async `aget` method for longer keys to allow this sleep.